scripts/tuple_tree_generator/tuple_tree_generator/templates/model_prelude.ts

```diff
@@ -0,0 +1,231 @@
+            return new Identifier(id);
+        } else {
+            return undefined;
+        }
+    }
+}
+
+const metaAddressTypes = [
+    "Invalid",
+    "Generic32",
+    "Generic64",
+    "Code_x86",
+    "Code_x86_64",
+    "Code_mips",
+    "Code_mipsel",
+    "Code_arm",
+    "Code_arm_thumb",
+    "Code_aarch64",
+    "Code_systemz",
+] as const;
+type MetaAddressType = typeof metaAddressTypes[number];
+
+const metaAddressRegex = new RegExp(
+    "^(?<Address>(0x[0-9a-fA-F]+)|)" +
+        `:(?<Type>${metaAddressTypes.join("|")})` +
+        "(:(?<Epoch>\\d+))?" +
+        "(:(?<AddressSpace>\\d+))?$"
+);
+
+function safeParseInt(str?: string): number | undefined {
+    if(str !== undefined) {
+        const num = Number.parseInt(str);
+        if(num == undefined || isNaN(num) || !isFinite(num)) {
+            return undefined;
+        } else {
+            return num;
+        }
+    } else {
+        return undefined;
+    }
+}
+
+export type IMetaAddress = string;
+
+export class MetaAddress {
+    address?: string;
+    type: MetaAddressType;
+    epoch?: number;
+    addressSpace?: number;
+
+    constructor(metaAddress: string) {
+        const match = metaAddress.match(metaAddressRegex);
+        if (match !== null && match.groups !== undefined) {
+            this.type = match.groups["Type"] as MetaAddressType;
+            this.address = match.groups["Address"];
+            this.epoch = safeParseInt(match.groups["Epoch"]);
+            this.addressSpace = safeParseInt(match.groups["AddressSpace"]);
+        } else {
+            this.type = "Invalid";
+        }
+    }
+
+    toJSON(): string {
+        function cleanStr(str?: string | number, end = true) {
+            if(str !== undefined) {
+                return end ? `:${str}` : str;
+            } else {
+                return ''
+            }
+        }
+        return `${cleanStr(this.address, false)}:${this.type}` +
+               `${cleanStr(this.epoch)}${cleanStr(this.addressSpace)}`;
+    }
+}
+
+export type IReference = string;
+
+export class Reference {
+    reference: string;
+
+    constructor(ref: string) {
+        this.reference = ref;
+    }
+
+    toJSON(): string | undefined {
+        return this.reference;
+    }
+
+    static parse(ref?: string): Reference | undefined {
+        if(ref !== undefined) {
+            return new Reference(ref);
+        } else {
+            return undefined;
+        }
+    }
+}
+
+function randomBytes(length: number): string {
+    let value = "";
+    for (let i = 0; i < length; i++) {
+        value += (Math.random() * 16).toString(16);
+    }
+    return value;
+}
+
+function randomBigint(start: bigint, end: bigint): bigint {
+    const diff = end - start;
+    for (;;) {
+        const randomLength = diff.toString().length * Math.random();
+        const value = BigInt(`0x${randomBytes(randomLength)}`);
+        if (value < diff) {
+            return value + start;
+        }
+    }
```

Comment:
This seem like a `do { ... } while` loop

Reply:
Fixed